### DISCUSSION TITLE
Opengl loaded but ktxTexture_GLUpload fails at glBindTexture

### DISCUSSION THREAD

**Discussion by viseztrance** [KhronosGroup/KTX-Software#707] [Q&A] [answered]
I'm using glad with glfw, and can call functions such as `glBindTexture(GL_TEXTURE_2D, textureID)` just fine. However calling `ktxTexture_GLUpload` fails, saying `Could not load OpenGL command: glBindTexture!`.

I looked over the source code, and it seems it's really calling gl.glBindTexture, but its implementation seems to be identical to what glad is already doing. Any ideas on what's going on?

Both glad and ktx are compiled statically.

**Comment by MarkCallow** [comment]
For a variety of excellent reasons libktx loads the GL function pointers dynamically instead of linking to libgl. The `gl` struct holds pointers to all the OpenGL functions used by `GLUpload`. The pointers are initialized the first time `GLUpload` is called. The message means it could not find the function or symbol `glBindTexture` when it did a `dlsym` (GNU/Linux, macOS) or `GetProcAddr` (Windows) on the OpenGL module. What OS are you using?

It must have found an OpenGL module in your app otherwise you would have seen the message `OpenGL lib not linked or loaded by application.` and it would not be running the code that prints the message you did see.

I do not know why it couldn't find glBindTexture. Are you building libktx yourself? If so, dId you make any changes? Is there a chance that the `#include "GL/glcorearb.h"` in glloader.c is somehow finding a modified glcorearb.h from GLAD?

**Reply by viseztrance** [reply]
Thank you for your fast and thorough reply. I'm on Fedora 38.

gcc --version
gcc (GCC) 13.1.1 20230511 (Red Hat 13.1.1-2)

I made no local changes to ktx. This is how I compiled it:

```sh
cmake .. -DCMAKE_CXX_FLAGS="-Wno-error=dangling-reference" -DCMAKE_C_FLAGS="-Wno-error=dangling-reference" \
-DKTX_FEATURE_STATIC_LIBRARY=ON -DCMAKE_BUILD_TYPE=Release -DKTX_FEATURE_VULKAN=OFF && \
make
```
"Wno-error=dangling-reference" is a fix for this https://github.com/fmtlib/fmt/issues/3415 (used by tools/imageio), basically an unrelated upstream gcc issue.

In addition to `vendor/KTX-Software/other_include/GL/glcorearb.h`, I do have another `glcorearb.h` file here, though I assume that's normal:
`/usr/include/GL/glcorearb.h`

I think I'll try to replicate this issue in a separate repository, and then change glad for glew etc.

**Comment by MarkCallow** [comment]
> I think I'll try to replicate this issue in a separate repository, and then change glad for glew etc.

Please do. I shall be happy to hear what you find out and happy to receive any PR or bug report. Note that `glBindTexture` is the first function in the list so it could be that the porblem has nothing to do with glBindTexture but is due to a more fundamental issue with the dlsym call retrieving GL function pointers. See Lines 98 - 106 in `lib/gl_funcs.c`.

**Reply by viseztrance** [reply, marked as answer]
As it turned out this was a silly mistake on my part, as I wasn't passing `-lGL` to the compile flag. It was rather annoying that it did work without it everywhere else, which made me jump to the wrong conclusions.
